### PR TITLE
Add support for listing parsers, bits and add Lp command

### LINKED ISSUE #17708: List parser plugins via e asm.parser=? and L subcommand
**Is your feature request related to a problem? Please describe.**

RParser plugins can be selected with the asm.parser eval variable, but there's no way to know which ones are registered, so its not possible to do autocompletion or listings.

**Describe the solution you'd like**

Do the same as in e asm.arch=?, but for asm.parser

Also i noticed that asm.bits=? is not working either.

**Describe alternatives you've considered**

Theres no other path

**Additional context**

```
$ r2 -
[0x00000000]> e asm.parser=?
^C
[0x00000000]>
[0x00000000]> L
Usage: L[acio]  [-name][ file]
| L             show this help
| L blah.dylib  load plugin file
| L-duk         unload core plugin by name
| Ll            list lang plugins (same as #!)
| LL            lock screen
| La            list asm/anal plugins (aL, e asm.arch=??)
| Lc            list core plugins
| Ld            list debug plugins (same as dL)
| LD            list supported decompilers (e cmd.pdc=?)
| Lm            list fs plugins (same as mL)
| Lh            list hash plugins (same as ph)
| Li            list bin plugins (same as iL)
| Lo            list io plugins (same as oL)
[0x00000000]> e asm.bits=?
[0x00000000]>
```

### PR DESCRIPTION
**Pull request by tantei3**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md#code-style)
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [radare2 book](https://github.com/radareorg/radare2book) with the relevant information (if needed)

**Detailed description**

For issue #17708 it is required to implement `e asm.parser=?` which should print all the parsers loaded currently. I've added code to update `node->options` list to have all the plugins listed.

Similarly `e asm.bits=?` is implemented by getting the value of `core->rasm->cur->bits` and populating all the `node->options` list to have all the available configurations.

Also the `Lp` command has been added to print all the parsers which will just call the `e asm.parser=?` command similar to implementation of `La`

Tab completion works for those fields now.
 
**Test plan**

Test case for `Lp` has been added to list the plugins (current output from my local terminal). I've tested locally for the changes made and autocomplete along with listing works for `asm.bits` and `asm.parser`.
**Closing issues**

closes #17708 
...

### CODE REVIEW

**Comment by trufae** [comment]
Cool! Thanks

**Comment by trufae** [comment]
Didnt tried the pr, but lgtm, is tab working in asm.parser after this?

**Comment by tantei3** [comment]
No it is still not working, that is why I mentioned it in the description. If you can point me to the implementation of tab function for like `e asm.arch=`, I can add it to both `bits` and `parser` hopefully

**Comment by tantei3** [comment]
Okay so I was able to find out that the implementation for the Tab autocompletion is the `find_e_opts` function and there, we always iterate over the `node->options` field. And it seems like for the `asm.bits` and `asm.parser` those options are not populated when they are created (haven't checked that part)

So one approach seems to be either in the `find_e_opts` write special checks to see if the string is `asm.bits` and `asm.parser` and manually duplicate the print logic over there. The other approach would seem to be like instantiating the node with the options while it is created and then autocomplete would work without changes to the that tab completion function. And the code to print would also probably be a bit more consistent and neat. 

I'll try to see if I can find out more about where node objects are created and if it is easy for me to populate the options field for them or not sometime later.

**Comment by tantei3** [comment]
The output from my local terminal, tab autocompletion works now with the same behaviour as previous commit.
```
$~/bin/radare2 -
 -- Wait a minute! I found a bug, self-fixing ... OK
[0x00000000]> e asm.parser=
6502.pseudo     arm.pseudo      att2intel       avr.pseudo      chip8.pseudo    dalvik.pseudo   m68k.pseudo     mips.pseudo
ppc.pseudo      sh.pseudo       tms320.pseudo   v850.pseudo     wasm.pseudo     x86.pseudo      z80.pseudo
[0x00000000]> e asm.parser=?
6502.pseudo
arm.pseudo
att2intel
avr.pseudo
chip8.pseudo
dalvik.pseudo
m68k.pseudo
mips.pseudo
ppc.pseudo
sh.pseudo
tms320.pseudo
v850.pseudo
wasm.pseudo
x86.pseudo
z80.pseudo
[0x00000000]> e asm.bits=
16        32        64
[0x00000000]> e asm.bits=?
16
32
64
[0x00000000]> e asm.arch=avr
[0x00000000]> e asm.bits=
8         16
[0x00000000]> e asm.bits=?
8
16
[0x00000000]> L?
Usage: L[acio]  [-name][ file]
| L          show this help
| L blah.so  load plugin file
| L-duk      unload core plugin by name
| Ll         list lang plugins (same as #!)
| LL         lock screen
| La         list asm/anal plugins (aL, e asm.arch=??)
| Lc         list core plugins
| Ld         list debug plugins (same as dL)
| LD         list supported decompilers (e cmd.pdc=?)
| Lm         list fs plugins (same as mL)
| Lh         list hash plugins (same as ph)
| Li         list bin plugins (same as iL)
| Lo         list io plugins (same as oL)
| Lp         list parser plugins (e asm.parser=?)
[0x00000000]> Lp
6502.pseudo
arm.pseudo
att2intel
avr.pseudo
chip8.pseudo
dalvik.pseudo
m68k.pseudo
mips.pseudo
ppc.pseudo
sh.pseudo
tms320.pseudo
v850.pseudo
wasm.pseudo
x86.pseudo
z80.pseudo
[0x00000000]>
```

File: libr/core/cconfig.c [hunk 1]
```diff
@@ -558,6 +558,21 @@ static void update_asmarch_options(RCore *core, RConfigNode *node) {
...
+static void update_asmbits_options(RCore *core, RConfigNode *node) {
+	if (core && core->rasm && core->rasm->cur && node) {
+		char buf[20];
+		int bits = core->rasm->cur->bits;
+		int i;
+		r_list_purge (node->options);
+		for (i = 1; i <= bits; i <<= 1) {
+			if (i & bits) {
+				snprintf (buf, 20, "%d", i);
+				SETOPTIONS (node, strdup (buf), NULL);
```
Comment (trufae) [inline libr/core/cconfig.c:570]: just use r_str_newf("%d", i);

File: libr/core/cconfig.c [hunk 2]
```diff
@@ -558,6 +558,21 @@ static void update_asmarch_options(RCore *core, RConfigNode *node) {
 	}
 }
 
+static void update_asmbits_options(RCore *core, RConfigNode *node) {
+	if (core && core->rasm && core->rasm->cur && node) {
+		char buf[20];
```
Comment (trufae) [inline libr/core/cconfig.c:563]: unnecessary

File: libr/core/cconfig.c [hunk 3]
```diff
@@ -558,6 +558,21 @@ static void update_asmarch_options(RCore *core, RConfigNode *node) {
...
+		int bits = core->rasm->cur->bits;
+		int i;
+		r_list_purge (node->options);
+		for (i = 1; i <= bits; i <<= 1) {
+			if (i & bits) {
+				snprintf (buf, 20, "%d", i);
```
Comment (trufae) [inline libr/core/cconfig.c:569]: remove this line, but if you need to pass fixed size buffers, is better to use sizeof than duplicuant constant values

**Review by trufae — CHANGES_REQUESTED** [review]
Fix the comment, add a test for the Lp command and LGTM!

File: libr/core/cconfig.c (hunk 1 above)
Comment (tantei3) [reply libr/core/cconfig.c]: Didn't know about that function! Updated the function accordingly.

File: libr/core/cconfig.c (hunk 2 above)
Comment (tantei3) [reply libr/core/cconfig.c]: Removed

File: libr/core/cconfig.c (hunk 3 above)
Comment (tantei3) [reply libr/core/cconfig.c]: Understood, I have replaced snprintf with r_str_newf.

File: test/db/cmd/cmd_list [hunk 4]
```diff
@@ -1,7 +1,7 @@
 NAME=List Parser Plugins Lp
 FILE=-
 CMDS=<<EOF
-Lp
+Lp | sort
```
Comment (trufae) [inline test/db/cmd/cmd_list:4]: Dont use system commands. You can sort with the ~ command too

**Review by trufae — CHANGES_REQUESTED** [review]
Or make the plugin list always sorted alphabetically when listing

**Comment by tantei3** [comment]
@trufae I did the change you suggested to use `~` to sort instead of `| sort` and tested in my local machine where it seemed to work. However in CI nothing is getting printed. I even copied the CI command `R2_NOPLUGINS=1 ./binr/radare2/radare2 -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'Lp ~' -` and ran it locally and it did print. So I am not sure how to get the correct output in CI.

**Comment by trufae** [comment]
The ~is just the cons filtering modifier. you have to specify what to do.

```
[0x100001060]> ~?? | grep sort
|  $[n]         sort numerically / alphabetically the Nth column
|  $!           sort in inverse order
[0x100001060]>
 ```

**Review by trufae — APPROVED** [review]
Looks perfect now!

**Comment by tantei3** [comment]
Thanks! I was confused because on my local machine it was printing without the `$[n]` and in the CI it wasn't.

File: test/db/cmd/cmd_list (hunk 4 above)
Comment (tantei3) [reply test/db/cmd/cmd_list]: Resolved

**Pull request merged**